libpod/container_internal_linux.go

```diff
@@ -1458,11 +1458,38 @@ func (c *Container) generateCurrentUserPasswdEntry() (string, error) {
 	if uid == 0 {
 		return "", nil
 	}
+
 	u, err := user.LookupId(strconv.Itoa(rootless.GetRootlessUID()))
 	if err != nil {
 		return "", errors.Wrapf(err, "failed to get current user")
 	}
-	return fmt.Sprintf("%s:x:%s:%s:%s:%s:/bin/sh\n", u.Username, u.Uid, u.Gid, u.Username, c.WorkingDir()), nil
+
+	// Lookup the user to see if it exists in the container image.
+	_, err = lookup.GetUser(c.state.Mountpoint, u.Username)
+	if err != nil && err != User.ErrNoPasswdEntries {
+		return "", err
+	}
+	if err == nil {
+		return "", nil
+	}
+
+	// If the user's actual home directory exists, or was mounted in - use
+	// that.
+	homeDir := c.WorkingDir()
+	hasHomedir := MountExists(c.config.Spec.Mounts, u.HomeDir)
+	if !hasHomedir {
+		ctrHomedir, err := securejoin.SecureJoin(c.state.Mountpoint, u.HomeDir)
+		if err == nil {
+			if s, err := os.Stat(ctrHomedir); err == nil && s.IsDir() {
+				hasHomedir = true
+			}
+		}
+	}
+	if hasHomedir {
+		homeDir = u.HomeDir
```

Comment:
Would it make sense to set WORKINGDIR here too (if unset)?
```console
$ ./bin/podman run -it --rm --userns=keep-id --security-opt label=disable -v /home/esm:/home/esm alpine_labels sh
/ $ pwd
/         <--- I kind of expected /home/esm
/ $ tail -1 /etc/passwd
esm:x:14904:14904:esm:/home/esm:/bin/sh     <--- it was detected
/ $ ls -la /home/esm
[as expected]                            <---- and it exists
```

Reply:
I'm reluctant to reset workdir given it could have been deliberately set by the user - I don't have a way of telling that this deep in the code.

Reply:
If workingdir == "" then we know it was never set.

Reply:
By this point, in the spec, it's already been defaulted to `/` if it was not set by the user.

Reply:
ok